From 8198c57b21d5e503f7e72221aa714aaabb2079cc Mon Sep 17 00:00:00 2001
From: Yuri Pankov <yuri.pankov@nexenta.com>
Date: Mon, 11 Dec 2017 10:11:25 +0300
Subject: [PATCH] OpenZFS 8897 - zpool online -e fails assertion when run on
 non-leaf vdevs
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Authored by: Yuri Pankov <yuri.pankov@nexenta.com>
Reviewed by: Toomas Soome <tsoome@me.com>
Reviewed by: Igor Kozhukhov <igor@dilos.org>
Reviewed-by: George Melikov <mail@gmelikov.ru>
Approved by: Dan McDonald <danmcd@joyent.com>
Ported-by: Brian Behlendorf <behlendorf1@llnl.gov>

OpenZFS-issue: https://www.illumos.org/issues/8897
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/9a551dd645
Closes #7030
---
 lib/libzfs/libzfs_pool.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

Index: zfs-linux-0.7.11/lib/libzfs/libzfs_pool.c
===================================================================
--- zfs-linux-0.7.11.orig/lib/libzfs/libzfs_pool.c
+++ zfs-linux-0.7.11/lib/libzfs/libzfs_pool.c
@@ -2510,6 +2510,7 @@ zpool_vdev_online(zpool_handle_t *zhp, c
 {
 	zfs_cmd_t zc = {"\0"};
 	char msg[1024];
+	char *pathname;
 	nvlist_t *tgt;
 	boolean_t avail_spare, l2cache, islog;
 	libzfs_handle_t *hdl = zhp->zpool_hdl;
@@ -2533,8 +2534,9 @@ zpool_vdev_online(zpool_handle_t *zhp, c
 	if (avail_spare)
 		return (zfs_error(hdl, EZFS_ISSPARE, msg));
 
-	if (flags & ZFS_ONLINE_EXPAND ||
-	    zpool_get_prop_int(zhp, ZPOOL_PROP_AUTOEXPAND, NULL)) {
+	if ((flags & ZFS_ONLINE_EXPAND ||
+	    zpool_get_prop_int(zhp, ZPOOL_PROP_AUTOEXPAND, NULL)) &&
+	    nvlist_lookup_string(tgt, ZPOOL_CONFIG_PATH, &pathname) == 0) {
 		uint64_t wholedisk = 0;
 
 		(void) nvlist_lookup_uint64(tgt, ZPOOL_CONFIG_WHOLE_DISK,
